model_analyzer/config/generate/neighborhood.py

```diff
@@ -0,0 +1,192 @@
+        return [Coordinate(x) for x in potential_values]
+
+    def _get_bounds(self, coordinate, radius):
+        bounds = []
+        for i in range(self._search_config.get_num_dimensions()):
+            dimension = self._search_config.get_dimension(i)
+
+            lower_bound = max(dimension.get_min_idx(), coordinate[i] - radius)
+            upper_bound = min(dimension.get_max_idx(),
+                              coordinate[i] + radius + 1)
+            bounds.append([lower_bound, upper_bound])
+        return bounds
+
+    def _enumerate_all_values_in_bounds(self, bounds):
+        possible_index_values = []
+        for bound in bounds:
+            possible_index_values.append(list(range(bound[0], bound[1])))
+
+        tuples = list(product(*possible_index_values))
+        return [list(x) for x in tuples]
+
+    def _get_unit_vector(self):
+        """
+        Create a unit vector pointing from the unweighted center of the
+        datapoints inside of the neighborhood to the weighted center of 
+        the datapoints inside of the neighborhood
+        """
+        coordinates, throughputs = self._compile_neighborhood_results()
+
+        coordinate_center = self._determine_coordinate_center(coordinates)
+        throughput_center = self._determine_weighted_coordinate_center(
+            coordinates, throughputs)
+
+        vector = throughput_center - coordinate_center
+
+        unit_vector = self._convert_to_unit_vector(vector)
+        return unit_vector
+
+    def _compile_neighborhood_results(self):
+        coordinates = []
+        results = []
+        for coordinate in self._neighborhood:
+            throughput = self._coordinate_data.get_throughput(coordinate)
+            if throughput is not None:
+                coordinates.append(deepcopy(coordinate))
+                results.append(throughput)
+        return coordinates, results
+
+    def _determine_coordinate_center(self, coordinates):
+        coordinate_center = Coordinate([0] *
+                                       self._search_config.get_num_dimensions())
+
+        for coordinate in coordinates:
+            coordinate_center += coordinate
+
+        num_coordinates = len(coordinates)
+        coordinate_center /= num_coordinates
+
+        return coordinate_center
+
+    def _determine_weighted_coordinate_center(self, coordinates, weights):
+        weighted_center = Coordinate([0] *
+                                     self._search_config.get_num_dimensions())
+
+        for i, _ in enumerate(weights):
+            weighted_center += coordinates[i] * weights[i]
+
+        weights_sum = sum(weights)
+
+        weighted_center /= weights_sum
+
+        return weighted_center
+
+    def _convert_to_unit_vector(self, vector):
+        magnitude = 0
+        for v in vector:
+            magnitude += math.pow(v, 2)
+        magnitude = math.sqrt(magnitude)
+
+        # Convert the vector to unit vector
+        if magnitude == 0:
+            # FIXME
```

Comment:
not quite sure what to do here. Original code returned None, and then handled it outside

Reply:
if magnitude is 0, then vector was all 0's.  I'll just return that vector.